File: hoffman-agents-python/src/conscious_agent/agent/world_state.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass, field
from collections.abc import Sequence


def sequence_to_state_id(sequence: Sequence[str]) -> int:
    data = "|".join(sequence)
    hash_bytes = hashlib.sha256(data.encode()).digest()
    return int.from_bytes(hash_bytes[:8], "big")


DIMENSION_BINS = {
    "temperature": 5,
    "intensity": 5,
    "valence": 5,
    "rhythm": 4,
}


@dataclass
class WorldState:
    sequences: dict[str, list[str]] = field(default_factory=dict)
    dimensions: dict[str, float] = field(default_factory=dict)

    def get_state_id(self) -> int:
        flat: list[str] = []
        for aid, seq in sorted(self.sequences.items()):
            flat.append(aid)
            flat.extend(seq)
        for k, v in sorted(self.dimensions.items()):
            n_bins = DIMENSION_BINS.get(k, 5)
            normalized = max(0.0, min(1.0, v))
            bin_idx = min(int(normalized * n_bins), n_bins - 1)
            flat.append(f"{k}:{bin_idx}")
        return sequence_to_state_id(flat)

    def dimension_delta(self, other: WorldState | None) -> dict[str, float]:
        if other is None:
            return {}
        delta = {}
        for k in self.dimensions:
            ov = other.dimensions.get(k, 0.0)
            delta[k] = self.dimensions[k] - ov
        return delta

    @staticmethod
    def from_sequence(agent_id: str, sequence: list[str]) -> WorldState:
        return WorldState(sequences={agent_id: sequence})

    def __bool__(self) -> bool:
        return len(self.sequences) > 0


EnvironmentState = WorldState

File: hoffman-agents-react-native/src/agent/world-state.js
const crypto = require('../crypto');

const DIMENSION_BINS = {
  temperature: 5, intensity: 5, valence: 5, rhythm: 4,
};

function sequenceToStateId(sequence) {
  const data = sequence.join('|');
  return crypto.createHash('sha256').update(data).digest().readUInt32BE(0);
}

class WorldState {
  constructor(sequences = {}, dimensions = {}) {
    this.sequences = sequences;
    this.dimensions = dimensions;
  }

  getStateId() {
    const flat = [];
    for (const [aid, seq] of Object.entries(this.sequences).sort()) {
      flat.push(aid, ...seq);
    }
    for (const [k, v] of Object.entries(this.dimensions).sort()) {
      const nBins = DIMENSION_BINS[k] || 5;
      const normalized = Math.max(0, Math.min(1, v));
      const binIdx = Math.min(Math.floor(normalized * nBins), nBins - 1);
      flat.push(`${k}:${binIdx}`);
    }
    return sequenceToStateId(flat);
  }

  dimensionDelta(other) {
    if (!other) return {};
    const delta = {};
    for (const k of Object.keys(this.dimensions)) {
      delta[k] = this.dimensions[k] - (other.dimensions[k] || 0);
    }
    return delta;
  }

  static fromSequence(agentId, sequence) {
    return new WorldState({ [agentId]: sequence });
  }
}

// WorldState is truthy if it has any sequences
WorldState.prototype[Symbol.toPrimitive] = function (hint) {
  if (hint === 'boolean') return Object.keys(this.sequences).length > 0;
  return null;
};

module.exports = { WorldState, EnvironmentState: WorldState, sequenceToStateId };
